client/src/pages/Chat.tsx

```diff
@@ -69,12 +66,54 @@ export const Chat = () => {
   }, [chat, user, navigate]);
 
   useEffect(() => {
-    // scroll to bottom of the chat when getting a new msg
+    // check if the scrolling element is in viewport
+    function isInViewport(element: HTMLElement) {
+      const rect = element.getBoundingClientRect();
+      return (
+        rect.top >= 0 &&
+        (rect.bottom - 50) <= (window.innerHeight || document.documentElement.clientHeight)
+      );
+    }
     if (scrollingRef.current) {
-      (scrollingRef.current as HTMLElement).scrollIntoView({ behavior: 'smooth' });
+      if (isInViewport(scrollingRef.current)) {
+        scrollToBottom('smooth');
+        setMsgDiff(0);
+      } else {
+        // if now in viewport - add button to scroll down
+        setMsgLength(prev => {
+          const diff = (messages?.length || 0) - prev;
+          setMsgDiff(prev => prev + diff);
+          return messages?.length || 0;
+        })
+      }
     }
   }, [messages]);
 
+  useEffect(() => {
+    // immediately scroll down when entering the page
+    if (scrollingRef.current) {
```

Comment:
It's good that you check if scrollingRef.current exists before attempting to scroll. This helps prevent errors if the ref is not yet initialized.